blog/content/posts/2025-12-xx-transport-direction.md

```diff
@@ -0,0 +1,87 @@
+
+We intend to make MCP stateless by:
+
+- Replacing the `initialize` handshake and sending the shared information with each request/response instead.
+- Providing a `discovery` mechanism for clients to query server capabilities if they need the information early, for scenarios such as UX hydration.
+
+With these changes, we're moving toward a more dynamic model where clients can optimistically "just try" what they need and receive clear error messages if a capability is unsupported.
+
+> **NOTE:** Many SDKs present a _`stateless`_ option in their Server transport configuration. This flag actually controls whether the `Mcp-Session-Id` header is used - read below for more on sessions.
+
+### Elevating Sessions
+
+Today, sessions are a side effect of the transport connection. With STDIO, sessions are implicit in the process lifecycle. With Streamable HTTP, sessions are created when a server assigns an `Mcp-Session-Id` during initialization. This conflates transport and application layer concerns.
+
+We plan to move sessions to the _data model layer_ - making them explicit rather than implicit.
+
+This means that MCP applications will be able to handle sessions as part of their domain logic. We are investigating the right way to do this - with a cookie style mechanism being the preferred choice for handling this state.
+
+Following the approach above makes MCP very similar to standard HTTP, where the protocol itself is stateless while applications build stateful semantics with cookies, tokens, and similar mechanisms. The exact approach to session creation is still being designed. Our goal is to remove existing ambiguities on what a session is in remote MCP scenarios.
+
+### Elicitations and Sampling
+
+Two existing MCP features enable key AI workflows that developers want: [Elicitations](https://spec.modelcontextprotocol.io/specification/2025-11-25/client/elicitation) for requesting human input, and [Sampling](https://spec.modelcontextprotocol.io/specification/2025-11-25/client/sampling) for agentic LLM interactions.
+
+To support these features at scale, the bidirectional communication pattern that these features rely on requires rethinking. Currently when a server needs more information to complete a tool call, it suspends operation and waits for a client response - meaning that it must remember all of its outstanding requests.
+
+To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The Server will return the Elicitation request as normal - however the client will return both request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message. This avoids managing potentially long-running state between specific nodes, and potentially removing the need for back-end storage.
```

Comment:
nit: 
```suggestion
To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The Server will return the Elicitation request as normal - however the client will return both request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message. This architectural shift eliminates the need for expensive 'sticky' load balancers or shared storage backends (like Redis) for simple ephemeral tool calls. Developers can build complex, multi-turn agentic behaviors on serverless platforms without necessarily managing persistent infrastructure.
```